odl/phantom/noise.py

```diff
@@ -105,6 +108,82 @@ def poisson_noise(intensity):
     return intensity.space.element(values)
 
 
+def salt_pepper_noise(vector, amount=0.05, salt_vs_pepper=0.5,
+                      low_val=None, high_val=None):
+    """Add salt and pepper noise to vector.
+
+    Salt and pepper noise replaces random elements in ``vector`` with
+    ``low_val`` or ``high_val``.
+
+    Parameters
+    ----------
+    vector : `FnBase` or `ProductSpace`
+        The vector that noise should be added to.
+    amount : float, optional
+        The propotion of elements in vector whose element should be converted
+        to noise.
+    salt_vs_pepper : float, optional
+        Relative aboundance of salt (high) vs pepper (low) noise. A high value
+        means more salt than pepper noise.
+    low_val : float, optional
+        The "pepper" color in the noise.
+        Default: minimum value of ``vector``. For product spaces, takes min for
+        each sub-space.
+    high_val : float, optional
+        The "salt" value in the noise.
+        Default: maximuim value of ``vector``. For product spaces, takes max
+        for each sub-space.
+
+    Returns
+    -------
+    salt_pepper_noise : ``vector.space`` element
+        ``vector`` with salt and pepper noise.
+
+    See Also
+    --------
+    white_noise
+    poisson_noise
+    """
+    from odl.space import ProductSpace
+
+    # Validate input parameters
+    amount, amount_in = float(amount), amount
+    if not (0 <= amount <= 1):
+        raise ValueError('`amount` ({}) should be a float in the interval '
+                         '[0, 1]'.format(amount_in))
+
+    salt_vs_pepper, salt_vs_pepper_in = float(salt_vs_pepper), salt_vs_pepper
+    if not (0 <= salt_vs_pepper <= 1):
+        raise ValueError('`salt_vs_pepper` ({}) should be a float in the '
+                         'interval [0, 1]'.format(salt_vs_pepper_in))
+
+    if isinstance(vector.space, ProductSpace):
+        values = [salt_pepper_noise(subintensity, amount, salt_vs_pepper,
+                                    low_val, high_val)
+                  for subintensity in vector]
+    else:
+        # Extract vector of values
+        values = as_flat_array(vector).copy()
```

Comment:
I want to access the vector as if it was flat, and asarray could return a square array